File: frontend/src/utils/polyhedron.ts
import earcut from "earcut"
import { convertToCartesian } from "./cartesian"

const parseCoords = (
  coords: number[],
  z: number | undefined,
  translateZ: number | undefined,
) => [
  coords[0],
  coords[1],
  (z !== undefined ? z : coords[2] || 0) + (translateZ || 0),
]

const parseRing = (
  ring: number[][],
  z: number | undefined,
  translateZ: number | undefined,
) => ring.map((coords: number[]) => parseCoords(coords, z, translateZ))

const sumUntilIndex = (arr: number[], index: number) =>
  arr
    // @ts-ignore
    .filter((value: number, idx: number) => idx < index)
    .reduce((sum: number, value: number) => sum + value, 0)

const triangulate = (coordinates: number[][][], reverse: boolean) => {
  const holes = coordinates
    // @ts-ignore
    .filter((coords, idx: number) => idx !== 0)
    .flatMap((c) => c.flat())
  const holesIndices = coordinates
    // @ts-ignore
    .map((coords, idx) =>
      coordinates[idx - 1] ? coordinates[idx - 1].length : null,
    )
    .filter((value) => value !== null)
    .map((value, idx, arr) => value + sumUntilIndex(arr, idx))
  const outer = coordinates[0].flat()
  const merged = [...outer, ...holes]
  const indices = earcut(merged, holesIndices, 3)
  const vertices = coordinates.flat()
  const polyhedron = []
  for (let i = 0; i < indices.length; i += 3) {
    polyhedron.push([
      convertToCartesian(vertices[indices[i]]),
      convertToCartesian(vertices[indices[i + (reverse ? 2 : 1)]]),
      convertToCartesian(vertices[indices[i + (reverse ? 1 : 2)]]),
      convertToCartesian(vertices[indices[i]]),
    ])
  }
  return polyhedron
}

const planeToWall = (lowerRing: number[][], upperRing: number[][]) => {
  const polyhedron = []
  for (let i = 0; i < lowerRing.length - 1; i++) {
    if (lowerRing[i + 1]) {
      const bl = convertToCartesian(lowerRing[i])
      const br = convertToCartesian(lowerRing[i + 1])
      const tl = convertToCartesian(upperRing[i])
      const tr = convertToCartesian(upperRing[i + 1])
      polyhedron.push([bl, tl, br, bl])
      polyhedron.push([br, tl, tr, br])
    }
  }
  return polyhedron
}

const generateWalls = (lower: number[][][], upper: number[][][]) => {
  // @ts-ignore
  return lower.flatMap((ring, idx) => {
    return planeToWall(lower[idx], upper[idx])
  })
}
const toPolyhedralSurface = (
  // @ts-ignore
  feature: any,
  [lower, upper]: number[][][][],
  removeBottomSurface: boolean | undefined,
) => {
  if (upper === undefined) {
    return triangulate(lower, false)
  }
  return [
    ...triangulate(upper, false),
    ...generateWalls(lower, upper),
    ...(removeBottomSurface ? [] : triangulate(lower, true)),
  ]
}

interface ParsersOptions {
  lowerLimit: number | undefined
  upperLimit: number | undefined
  translateZ: number | undefined
  removeBottomSurface: boolean | undefined
}

const parsers: any = {
  Polygon: (feature: any, options: ParsersOptions) => {
    const { lowerLimit, upperLimit, translateZ, removeBottomSurface } =
      options || {}
    if (lowerLimit === undefined && upperLimit === undefined) {
      const lower = feature.geometry.coordinates.map((ring: number[][]) =>
        parseRing(ring, lowerLimit, translateZ),
      )
      return toPolyhedralSurface(feature, [lower], removeBottomSurface)
    }
    const lower = feature.geometry.coordinates.map((ring: number[][]) =>
      parseRing(ring, lowerLimit, translateZ),
    )
    const upper = feature.geometry.coordinates.map((ring: number[][]) =>
      parseRing(ring, upperLimit, translateZ),
    )
    const averageZLower =
      lower[0].reduce((sum: number, coords: number[]) => sum + coords[2], 0) /
      lower[0].length
    const averageZUpper =
      upper[0].reduce((sum: number, coords: number[]) => sum + coords[2], 0) /
      upper[0].length
    return toPolyhedralSurface(
      feature,
      averageZLower > averageZUpper ? [upper, lower] : [lower, upper],
      removeBottomSurface,
    )
  },
  MultiPolygon: (feature: any, options: ParsersOptions) => {
    return feature.geometry.coordinates.flatMap(
      (coordinates: number[][][][]) => {
        return parsers.Polygon(
          { ...feature, geometry: { type: "Polygon", coordinates } },
          options,
        )
      },
    )
  },
}

export const collectionToPolyhedralSurfaceZ = (
  collection: any,
  {
    filter = (feature: any) => feature,
    // @ts-ignore
    computeOptions = (feature: any) => ({
      lowerLimit: undefined,
      upperLimit: undefined,
      translateZ: undefined,
      removeBottomSurface: true,
    }),
  } = {},
) => {
  if (!collection?.features) {
    return {
      type: "FeatureCollection",
      features: [],
    }
  }
  const features = collection.features.filter(filter).map((feature: any) => {
    const options = computeOptions(feature)
    const coordinates = parsers[feature.geometry.type](feature, options)
    return {
      ...feature,
      geometry: {
        type: "POLYHEDRALSURFACE Z",
        coordinates,
      },
    }
  })
  return {
    type: "FeatureCollection",
    features,
  }
}

File: backend/app/worker/tasks/pg2b3dm/polyhedron.py
from app.worker.tasks.pg2b3dm.processes import earcut


def parse_coords(coords, z, translate_z):
    height = coords[2] if coords[2] else 0
    if z:
        height = z
    _translate_z = translate_z if translate_z else 0
    return [coords[0], coords[1], height + _translate_z]


def parse_ring(ring, z, translate_z):
    return [parse_coords(coords, z, translate_z) for coords in ring]


def sum_until_index(arr, index):
    return sum(arr[:index])


def triangulate(coordinates, reverse):
    indices = earcut(coordinates)

    vertices = []
    for rings in coordinates:
        for ring in rings:
            vertices.append(ring)

    polyhedron = []
    for i in range(len(indices)):
        if i % 3 == 0:
            if reverse:
                polyhedron.append([
                    vertices[indices[i]],
                    vertices[indices[i + 2]],
                    vertices[indices[i + 1]],
                    vertices[indices[i]]
                ])
            else:
                polyhedron.append([
                    vertices[indices[i]],
                    vertices[indices[i + 1]],
                    vertices[indices[i + 2]],
                    vertices[indices[i]]
                ])
    return polyhedron


def plane_to_wall(lower_ring, upper_ring):
    polyhedron = []
    for i in range(len(lower_ring) - 1):
        if lower_ring[i + 1]:
            bl = lower_ring[i]
            br = lower_ring[i + 1]
            tl = upper_ring[i]
            tr = upper_ring[i + 1]
            polyhedron.append([bl, tl, br, bl])
            polyhedron.append([br, tl, tr, br])
    return polyhedron


def generate_walls(lower, upper):
    walls = []
    for idx in range(len(lower)):
        walls += plane_to_wall(lower[idx], upper[idx])
    return walls


def to_polyhedral_surface(lower, upper, remove_bottom_surface):
    if not upper:
        return triangulate(lower, False)

    polyhedral_surface = []
    polyhedral_surface += triangulate(upper, False)
    polyhedral_surface += generate_walls(lower, upper)

    if not remove_bottom_surface:
        polyhedral_surface += triangulate(lower, True)

    return polyhedral_surface


def polygon_to_polyhedral_surface(geometry, options):
    lower_limit = options['lower_limit']
    upper_limit = options['upper_limit']
    translate_z = options['translate_z']
    remove_bottom_surface = options['remove_bottom_surface']

    lower = [parse_ring(ring, lower_limit, translate_z) for ring in geometry['coordinates']]

    if lower_limit is None and upper_limit is None:
        return to_polyhedral_surface(lower, None, remove_bottom_surface)

    upper = [parse_ring(ring, upper_limit, translate_z) for ring in geometry['coordinates']]

    average_z_lower = sum(c[2] for c in lower[0]) / len(lower[0])
    average_z_upper = sum(c[2] for c in upper[0]) / len(upper[0])

    if average_z_lower > average_z_upper:
        return to_polyhedral_surface(upper, lower, remove_bottom_surface)
    return to_polyhedral_surface(lower, upper, remove_bottom_surface)


def geometry_to_polyhedral_surface(geometry, options):
    if geometry['type'] == 'Polygon':
        return polygon_to_polyhedral_surface(geometry, options)
    return []


def polyhedral_to_wkt(polyhedron):
    triangles = []
    for triangle in polyhedron:
        vertices = []
        for vertex in triangle:
            vertices.append(" ".join(map(str, vertex)))
        triangles.append(f"(({', '.join(vertices)}))")
    return f"POLYHEDRALSURFACE Z({','.join(triangles)})"
